crates/net/network/src/receiver.rs

```diff
@@ -0,0 +1,32 @@
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::Future;
+use tokio::sync::oneshot::{error::RecvError, Receiver};
+
+pub struct ResolvedOneshotReceiver<T> {
+    receiver: Receiver<T>,
```

Comment:
you can use `pin_project` here and annotate the receiver with `#[pin]`, scan other parts of the code base for reference.